src/crafting.h

```diff
@@ -104,6 +104,84 @@ struct recipe {
 
 };
 
+/**
+*   enum used by comp_selection to indicate where a component should be consumed from.
+*/
+enum usage {
+    use_from_map = 1,
+    use_from_player = 2,
+    use_from_both = 1 | 2,
+    use_from_none = 4,
+    cancel = 5 // FIXME: hacky.
+};
+
+/**
+*   Struct that represents a selection of a component for crafting.
+*/
+template<typename CompType = component>
+struct comp_selection {
+    /** Tells us where the selected component should be used from. */
+    usage use_from = use_from_none;
+    CompType comp;
+
+    /** provides a translated name for 'comp', suffixed with it's location e.g '(nearby)'. */
+    std::string nname() const;
+};
+using item_selection = comp_selection<item_comp>;
```

Comment:
I know this contradicts what BevapDin said, sorry.
I'm really not a fan of the using directive (or its antecedent typedef).  Compound types should either be spelled out when used, or properly encapsulated in a class definition. Aliases like this make the code hard to read, because there are more places you have to read to determine what a particular piece of code means.